File: core/src/commonMain/kotlin/work/socialhub/kmastodon/entity/share/RateLimit.java
package work.socialhub.kmastodon.entity.share;

import net.socialhub.http.HttpResponse;
import net.socialhub.logger.Logger;

import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.TimeZone;

/**
 * @author uakihir0
 */
public class RateLimit {

    private static final Logger logger = Logger.getLogger(RateLimit.class);

    private static final String X_RATELIMIT_LIMIT = "X-RateLimit-Limit";

    private static final String X_RATELIMIT_REMAINING = "X-RateLimit-Remaining";

    private static final String X_RATELIMIT_RESET = "X-RateLimit-Reset";

    private int limit;
    private int remaining;
    private Date reset;

    public static RateLimit of(HttpResponse response) {
        if (response == null) {
            return null;
        }

        try {
            String limit = response.getResponseHeader(X_RATELIMIT_LIMIT);
            String remaining = response.getResponseHeader(X_RATELIMIT_REMAINING);
            String reset = response.getResponseHeader(X_RATELIMIT_RESET);

            RateLimit rateLimit = new RateLimit();
            rateLimit.setLimit(new Integer(limit));
            rateLimit.setRemaining(new Integer(remaining));

            String[] dates = reset.split("\\.");
            DateFormat format = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss");
            format.setTimeZone(TimeZone.getTimeZone("UTC"));
            rateLimit.setReset(format.parse(dates[0]));

            return rateLimit;

        } catch (Exception e) {
            logger.debug(e.getMessage());
            return null;
        }
    }

    public int getLimit() {
        return limit;
    }

    public void setLimit(int limit) {
        this.limit = limit;
    }

    public int getRemaining() {
        return remaining;
    }

    public void setRemaining(int remaining) {
        this.remaining = remaining;
    }

    public Date getReset() {
        return reset;
    }

    public void setReset(Date reset) {
        this.reset = reset;
    }
}

File: core/src/commonMain/kotlin/work/socialhub/kmastodon/entity/share/RateLimit.kt
package work.socialhub.kmastodon.entity.share

import kotlinx.datetime.Instant
import kotlinx.datetime.toInstant
import work.socialhub.khttpclient.HttpResponse
import kotlin.js.JsExport

@JsExport
class RateLimit {
    var limit: Int = 0
    var remaining: Int = 0

    @Suppress("NON_EXPORTABLE_TYPE")
    var reset: Instant? = null

    companion object {
        private const val X_RATELIMIT_LIMIT = "X-RateLimit-Limit"
        private const val X_RATELIMIT_REMAINING = "X-RateLimit-Remaining"
        private const val X_RATELIMIT_RESET = "X-RateLimit-Reset"

        @Suppress("NON_EXPORTABLE_TYPE")
        fun of(response: HttpResponse): RateLimit? {

            try {
                val limit = response.headers[X_RATELIMIT_LIMIT]
                val remaining = response.headers[X_RATELIMIT_REMAINING]
                val reset = response.headers[X_RATELIMIT_RESET]

                return RateLimit().also {
                    if (limit?.isNotEmpty() == true) {
                        it.limit = limit[0].toInt()
                    }
                    if (remaining?.isNotEmpty() == true) {
                        it.remaining = remaining[0].toInt()
                    }
                    if (reset?.isNotEmpty() == true) {
                        it.reset = reset[0].toInstant()
                    }
                }
            } catch (e: Exception) {
                return null
            }
        }
    }
}
